### ISSUE TITLE
--variations and --shaper=directwrite

### ISSUE THREAD

**Issue by simoncozens** [harfbuzz/harfbuzz#3066]
CoreText *does* support FeatureVariations (I checked with a HTML demo in Safari). But it seems like `hb-shape --shaper=coretext` does not:

```
$ hb-shape bahnschrift.ttf '$' --variations=wght=100 --shaper=ot
[gid696=0+1181]
 hb-shape bahnschrift.ttf '$' --variations=wght=1000 --shaper=ot
[gid730=0+1181]

$ hb-shape bahnschrift.ttf '$' --variations=wght=100 --shaper=coretext
[gid730=0+1181]
$ hb-shape bahnschrift.ttf '$' --variations=wght=1000 --shaper=coretext
[gid730=0+1181]
```

**Comment by behdad** [comment]
We don't call CoreText variations APIs. Any idea what the API is called? cc @drott

**Comment by khaledhosny** [comment]
This is probably true for `directwrite` shaper (and `uninscribe`, but I don’t think GDI has a variations API).

**Comment by simoncozens** [comment]
Yes, directwrite doesn't work either (but then DW doesn't [really support variations](https://twitter.com/simoncozens/status/1418162669069246465) too well anyway...). An example of the DW API can be found [here](https://github.com/microsoft/DWriteShapePy/blob/16b7ea0990e94af342598ab28601cfc1b9dc3f3a/src/cpp/DWriteShapeLib.cpp#L300-L313).

**Comment by drott** [comment]
> We don't call CoreText variations APIs. Any idea what the API is called? cc @drott @jfkthame

You mean what API is needed for applying variations to a `CTFontRef`? I think that would be [CTFontCreateCopyWithAttributes](https://developer.apple.com/documentation/coretext/1511225-ctfontcreatecopywithattributes?language=objc) - and specifying the variations for the `kCTFontVariationAttribute` key of the attributes dictionary, example [here in Skia](https://source.chromium.org/chromium/chromium/src/+/main:third_party/skia/src/ports/SkTypeface_mac_ct.cpp;l=1215).

> Yes, directwrite doesn't work either (but then DW doesn't [really support variations](https://twitter.com/simoncozens/status/1418162669069246465) too well anyway...). An example of the DW API can be found [here](https://github.com/microsoft/DWriteShapePy/blob/16b7ea0990e94af342598ab28601cfc1b9dc3f3a/src/cpp/DWriteShapeLib.cpp#L300-L313).

The Windows 10 Creators Update did only have limited support - that's correct, however, subsequent Windows versions did deliver more complete variations support. In Chrome we do use DWrite for variable fonts when sufficient support is available. See for example the Skia code that maps `SkTypeface` variation information to DWrite in [SkTypeface_win_dw.cpp](https://source.chromium.org/chromium/chromium/src/+/main:third_party/skia/src/ports/SkTypeface_win_dw.cpp;l=316).

**Comment by jfkthame** [comment]
On 22/07/2021 22:08, Simon Cozens wrote:
> Yes, directwrite doesn't work either (but then DW doesn't really support 
> variations <https://twitter.com/simoncozens/status/1418162669069246465> 
> too well anyway...).

The doc shown in that tweet refers to the Windows 10 Creators Update 
release, from April 2017. Things have moved on considerably since then.

FWIW, in Firefox we consider variation fonts to be "supported" starting 
with the Fall Creators Update (build 16299).

JK

**Comment by behdad** [comment]
For DWrite, also: https://github.com/microsoft/DWriteShapePy/blob/main/src/cpp/DWriteShapeLib.cpp#L300
